### PR TITLE
Case-insensitive language tags

### LINKED ISSUE #2039: Normalized language tags.
In RDF 1.1, it says that language tags are compared case insensitively and also that "Lexical representations of language tags  MAY be converted to lower case." It is in RDF semantics as D-entailment.

The issue has come to prominence because of work on RDF canonicalization and hashing (RCH) which works on the syntax of graphs. Signing and Verifiable Credentials than rely on RCH. So syntax matters, not the value.

Jena5 will switch to normalized language tags. e.g. "en-GB"
Data should be reloaded to be safe.

https://lists.apache.org/thread/l6trf46ko83vdg3hfzgrvbhgcotzv4xc (2023-09-28)

NB Canonicalization of language tags is different - that involves , for example, replacing the country code with the preferred code.

The correct termninology from BCP47/RFC5646 is "format".

### LINKED ISSUE #2051: Support for initial text direction (RDF 1.2)
### Change

https://www.w3.org/TR/rdf12-concepts/#section-text-direction

https://www.w3.org/TR/rdf12-n-triples/#grammar-production-LANG_DIR

Language literals can be `"abc"@en--ltr` (left-to-right). 

This issue is to add the RDF term support, not provide in every syntax form because if #2039 (normalized language tags) is happening, it makes sense to do at the same time because changes in the same area.

### Are you interested in contributing a pull request for this task?

Yes

### LINKED ISSUE #2126: Clearer naming in NodeFactory for string literal creation
### Change

There are many "createLiteral" functions.

Migrate to `createLiteralString`, `createLiteralLang` and `createLiteralDirLang` for the variations on string literals.

### Are you interested in contributing a pull request for this task?

Yes

### PR DESCRIPTION
**Pull request by afs**
GitHub issue resolved #2039 
This closes #2126
This closes #2051

This PR makes language tags case insensitive.  This is mandated in RDF 1.2. In RDF 1.1, and before, language tags only had to be treated as case-insensitive values.

Internally, language tags are formatted using the light-weight algorithm in [RFC 5646, section 2.1.1](https://datatracker.ietf.org/doc/html/rfc5646#section-2.1.1). It does not validate language tags.

`"abc"@en` and `"abc"@EN` become the same RDF term.

`LANG` returns formatted language tags.
e.g. `en`, `en-GB`, `en-Latn-GB`

The PR also includes the node machinery for initial text direction (RDF 1.2). It does not include API or SPARQL support.

Language tag validation during parsing may be added later.

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx, or if in JIRA, JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

### CODE REVIEW

**Comment by afs** [comment]
This PR is my last planned Jena4/Jena5 incompatible change.

**Comment by namedgraph** [comment]
Merry Christmas :)

**Comment by afs** [comment]
Replace uses of `toLoweCase(Locale.ROOT)` with calls to a common library functions for "system " uses. This is common in the language tag related code. Ditto `toUpperCase`.

File: jena-core/src/main/java/org/apache/jena/graph/impl/LiteralLabel.java [hunk 1]
```diff
@@ -451,61 +479,100 @@ public boolean sameValueAs( LiteralLabel other ) {
...
+    /**
+     * Return true if the literal label is a well-formed language string with text direction.
+     * This excludes "abc"^^rdf:dirLangString.
+     */
+    private static boolean isLangStringDir(LiteralLabel lit) {
+        // Assume well formed.
+        String lang = lit.language();
+        // Allow "abc"@--rtl
+//        if ( isEmpty(lit.language()) )
+//            return false;
```
Comment (kinow) [inline jena-core/src/main/java/org/apache/jena/graph/impl/LiteralLabel.java:569]: Interesting! :point_up:

**Review by kinow — APPROVED** [review]
Minor comments, no blockers. +1

File: jena-arq/src/test/java/org/apache/jena/sparql/expr/TestExpressions.java [hunk 2]
```diff
@@ -346,9 +346,29 @@ public class TestExpressions
...
     @Test public void boolean_128() { testBoolean("datatype('fred'^^<foo>) = <Foo>", false) ; }
-    @Test public void string_15() { testString("lang('fred'@en)", "en") ; }
-    @Test public void string_16() { testString("lang('fred'@en-uk)", "en-uk") ; }
-    @Test public void string_17() { testString("lang('fred')", "") ; }
+
+    @Test public void lang_01() { testString("LANG('tea time'@en)", "en") ; }
+    // Aside For some strange reason, the language code is GB not UK.
+    // The state is UK! "The United Kingdom of Great Britain and Norther Ireland."
+    // The four countries England, Scotland, Wales and Northern Ireland (since 1922).
+    // It's complicated: https://en.wikipedia.org/wiki/United_Kingdom
```
[the remark this answers is not in the document]
Comment (afs) [reply jena-arq/src/test/java/org/apache/jena/sparql/expr/TestExpressions.java:354]: Yes, it's `en-GB`. 

Digression: London's special economy is not a legal structure but reflects the way investment funding flows in the UK.

File: jena-base/src/main/java/org/apache/jena/atlas/lib/Lib.java [hunk 3]
```diff
@@ -112,25 +113,42 @@ static public UnsupportedOperationException unsupportedMethod(Object object, Str
...
+    /** Non-locale lowercase {@link Locale#ROOT} */
+    public static String lowercase(String string) {
+        // Hide implementation
+        return string.toLowerCase(Locale.ROOT);
+    }
+
+    /** Non-locale uppercase {@link Locale#ROOT} */
+    public static String uppercase(String string) {
+        // Hide implementation
+        return string.toUpperCase(Locale.ROOT);
```
[the remark this answers is not in the document]
Comment (afs) [reply jena-base/src/main/java/org/apache/jena/atlas/lib/Lib.java:146]: This is a specific form of uppercase for "system" use - there points in the system where lower/uppercase apply to 
(e.g. language tags and URI schema name which are `a-zA-Z`). There are some locales where lower/upper case in the locale do not map ASCII to ASCII (lowercase dotted `i` - some languages have uppercase dotted `I`).

File: jena-arq/src/main/java/org/apache/jena/sparql/util/StringUtils.java [hunk 4]
```diff
@@ -48,28 +48,28 @@ public static String str(float value)
...
-    
+
     public static <T> String str(T[] array)
     {
         return Arrays.asList(array).toString() ;
     }
 
-    private static Pattern p = Pattern.compile("http:[^ \n]*[#/]([^/ \n]*)") ;
-    /** Abbreviate, crudely, URI in strings, leaving only their last component. */ 
+    private static Pattern p = Pattern.compile("https?:[^ \n]*[#/]([^/ \n]*)") ;
```
[the remark this answers is not in the document]
Comment (afs) [reply jena-arq/src/main/java/org/apache/jena/sparql/util/StringUtils.java:62]: `https?` matches `http://` or `https://` - the `?` applies to the `s` only.

File: jena-arq/src/test/java/org/apache/jena/riot/TestRDFParser.java [hunk 5]
```diff
@@ -231,17 +231,17 @@ private RDFParserBuilder builder() {
...
+//        testNormalization("'abc'@En-gB", "'abc'@En-gB", builder().langTagAsGiven());
+//    }
+//
+//    @Test public void canonical_langTag_2() {
+//        testNormalization("'abc'@En-gB", "'abc'@en-gb", builder().langTagLowerCase());
+//    }
+//
+//    @Test public void canonical_langTag_3() {
+//        testNormalization("'abc'@En-gB", "'abc'@en-GB", builder().langTagCanonical());
+//    }
```
[the remark this answers is not in the document]
Comment (afs) [reply jena-arq/src/test/java/org/apache/jena/riot/TestRDFParser.java:244]: Added a comment "// Old tests that applied to Java4. Delete in the future."

Leaving the commented tests record the jena4 world.

File: jena-arq/src/main/java/org/apache/jena/sparql/util/StringUtils.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kinow) [reply jena-arq/src/main/java/org/apache/jena/sparql/util/StringUtils.java:62]: Aaaahh! Hadn't seen the `?`! :laughing: sorry!

File: jena-arq/src/test/java/org/apache/jena/riot/TestRDFParser.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kinow) [reply jena-arq/src/test/java/org/apache/jena/riot/TestRDFParser.java:245]: :+1: +1! I saw another set of tests with a useful comment, thanks!

File: jena-base/src/main/java/org/apache/jena/atlas/lib/Lib.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kinow) [reply jena-base/src/main/java/org/apache/jena/atlas/lib/Lib.java:146]: > for "system" use

:+1: then I think this is fine, marking as resolved! Thanks!

File: jena-core/src/main/java/org/apache/jena/datatypes/xsd/impl/RDFDirLangString.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * rdf:dirLangString (literal with language and initial text direction)
+ * This covers the unusual case of "foo"^^rdf:langString or "foo"^^rdf:dirLangString.
+ * When there is a language tag, there is a lexical form but it is in two parts lex@lang or lex@lang--ltr
+ */
+
+public class RDFDirLangString extends BaseDatatype implements RDFDatatype {
+
+    /** Singleton instance */
+    // Include the string for the RDF namespace, not use RDF.getURI(), to avoid an initializer circularity
```
[the remark this answers is not in the document]
Comment (afs) [reply jena-core/src/main/java/org/apache/jena/datatypes/xsd/impl/RDFDirLangString.java]: Done

**Pull request merged**